File: solutions/python/matching-brackets/1/matching_brackets.py
dict = {
    "[": "]",
    "{": "}",
    "(": ")"
}
def is_paired(input_string):
    pairs = []
    for letter in input_string:
        if letter in dict:
            pairs.append(letter)
            if input_string.count(letter) != input_string.count(dict[letter]):
                return False
        elif letter in dict.values() and len(pairs) > 0:
            last_item = pairs[-1]
            if last_item in dict and dict[last_item] == letter:
                pairs.pop()
            else:
                return False
    return len(pairs) == 0

File: solutions/csharp/matching-brackets/1/MatchingBrackets.cs
public static class MatchingBrackets {
  private const string OpeningBrackets = "([{";
  private const string ClosingBrackets = ")]}";

  public static bool IsPaired(string input) {
    var stack = new Stack<char>();
    foreach (var character in input) {
      if (OpeningBrackets.Contains(character)) {
        stack.Push(character);
      }
      else if (ClosingBrackets.Contains(character)) {
        if (stack.Count == 0) {
          return false;
        }

        var lastOpening = stack.Pop();
        if (OpeningBrackets.IndexOf(lastOpening) != ClosingBrackets.IndexOf(character)) {
          return false;
        }
      }
    }

    return !stack.Any();
  }
}
